### DISCUSSION TITLE
Is this expected functionality for `Example` objects?

### DISCUSSION THREAD

**Discussion by mikeyhogarth** [explosion/spaCy#7817] [Help: Coding & Implementations] [answered]
I'm a big fan of Spacy and this week decided to check out version 3.0 by doing an `EntityLinker` experiment. I ran straight into issues around the `Example` objects - I think I've found a bug but I don't think I really have a deep enough grasp of spacy to know for sure.

The effect I was seeing is that the entity linker, when I started training it, was complaining about a `sentencizer` not existing in the pipeline. I tried adding one and it didint' fix the issue. I did a bit of digging and eventually managed to boil it down to being casued by a line referenceing the `reference` within the `Example`, so easily reproducible.

I wanted to ask the community whether or not the following is the intended functionality;

```python
import spacy
from spacy.training import Example

# Create nlp function and add sentencizer to pipeline
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("sentencizer")

# Create a document and build an example from it
doc = nlp("Bono is the singer of U2")
example = Example.from_dict(doc, {"links": {(0, 4): { "Q834621": 1.0 }}})

# Print out the sentences from the example's reference 
for sent in example.reference.sents:
    print(sent)

# ---------------------------------------------------------------------------
# ValueError                                Traceback (most recent call last)
# <ipython-input-17-7fc2e1155fc5> in <module>
#      11 
#      12 # Print out the sentences from the example's reference
# ---> 13 for sent in example.reference.sents:
#      14     print(sent)
# /opt/conda/lib/python3.8/site-packages/spacy/tokens/doc.pyx in sents()

# ValueError: [E030] Sentence boundaries unset. You can add the 'sentencizer' component to the pipeline with:
#  `nlp.add_pipe('sentencizer')`. Alternatively, add the dependency parser or sentence recognizer, or set sentence 
# boundaries by setting `doc[i].is_sent_start`.
```

My understanding is that the `reference` on the example is supposed to be whatever you passed into `from_dict` but with the reference annotations added on from the `example_dict` I passed in (the one with the links). I've looked through the Emerson example that Sophie did but that didn't get me the answers because she's using prodigy and gets a load more annotation than just the links, so isnt' really comparable with what I'm doing here where I just want to provide the `links` annotation and nothing else.

Anyway hopefully this makes sense! Apologies if I've missed something really obvious. 

If this does turn out to be a bug I am open to helping to fix / doing you a PR if it would help.

Cheers!

**Comment by adrianeboyd** [comment, marked as answer]
Yes, it's expected that the `example.reference` from the following code wouldn't have sentence boundaries:

```python
example = Example.from_dict(doc, {"links": {(0, 4): { "Q834621": 1.0 }}})
```

The `reference` doc includes the annotation provided in the dict, using the tokenization from the `doc` if you don't provide `words`.

The `predicted` doc is the provided `doc` arg.

To modify the reference doc, you can add `sent_starts` to the dict of annotations or you can just apply the sentencizer to add them to the reference doc:

```python
example.reference = nlp.get_pipe("sentencizer")(example.reference)
```

This is the nice thing about working directly with `Doc` objects instead of `GoldParse` from v2: you can add anything to the reference doc in the exact same way you'd add it to the predicted doc in a working pipeline.

**Reply by mikeyhogarth** [reply]
Thanks for the response - based on that I still think there's an issue (or possibly something that I am just confused about) - I'm looking at the example here; 
https://spacy.io/api/data-formats

```python
# Training data for an Entity Linking component
doc = nlp("Russ Cochran his reprints include EC Comics.")
gold_dict = {"links": {(0, 12): {"Q7381115": 1.0, "Q2146908": 0.0}}}
example = Example.from_dict(doc, gold_dict)
```

This suggests you should be able to create an example for training an `EntityLinker` purely from the above.

However, the entity linker requires that entities exist on the reference document (as per https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L228)

When I try and add them in as per the recommendation above though, it blats the links;

```python
import spacy
from spacy.training import Example
nlp = spacy.load("en_core_web_sm")

# Create a document and build an example from it
doc = nlp("Bono is the singer of U2")
example = Example.from_dict(doc, {"links": {(0, 4): { "Q834621": 1.0 }}})

# Check ents and kb_ids...
print(example.reference.ents)
print(example.get_aligned("ENT_KB_ID", as_string=True))

# Output as follows:
#  ()
#  ['Q834621', '', '', '', '', '']

# We need the reference document to have entities or the entity linker will fail, so run it through the NER pipe...
example.reference = nlp.get_pipe("ner")(example.reference)

# Now we get our entities, but the links dissapear!
print(example.reference.ents)
print(example.get_aligned("ENT_KB_ID", as_string=True))

# Output as follows - we do now have entities but we've lost the KB_IDs :(
# (Bono,)
# ['', '', '', '', '', '']
```
I am beginning to think that it's not possible to create training examples for an entity linker purely as per the Russ Cochran example given in the docs - am I missing something or is that indeed the case?

**Reply by mikeyhogarth** [reply]
This is what I've eventually ended up with , just wanted to make sure i wasn't doing anything crazy!

```python
import spacy
from spacy.training import Example
nlp = spacy.load("en_core_web_sm")
nlp.add_pipe("sentencizer")

# Create a document and annotation dict
doc = nlp("Bono is the singer of U2")
annot = {"links": {(0, 4): { "Q834621": 1.0 }}}

# Add the entities into the annotation by pinching what the model already parsed...
annot["entities"] = [(ent.start_char, ent.end_char, ent.label_) for ent in doc.ents]
example = Example.from_dict(doc, annot)

# Finally add in the sentencizer because it's needed by the entity linker
example.reference = nlp.get_pipe("sentencizer")(example.reference)

# Check things are okay
print(example.reference.ents)
print(example.get_aligned("ENT_KB_ID", as_string=True))

# Output as follows - success!
# (Bono,)
#  ['Q834621', '', '', '', '', '']
```
And it now works when I train it - woo!

I think it might be worth putting a little note in the docs by that Russ Cochran example though about needing entities to actually train the linker (unless I've grabbed completely the wrong end of the stick - in which case my apologies!).

**Reply by svlandeg** [reply]
Hi @mikeyhogarth , 

yes, you're right, the current documentation is confusing, because the Entity Linker requires gold sentences and entities. I've clarified this in the docs in this PR: https://github.com/explosion/spaCy/pull/7839

**Comment by svlandeg** [comment]
Just want to add to this thread, that we've updated [this notebook](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/notebooks/notebook_video.ipynb) with the full run-through on how to create and train an entity linker in spaCy v3. The notebook shows the code if you want to implement/run the training loop in code, and also runs the `sentencizer` on the `Example` `reference` objects.

**Comment by narayanacharya6** [comment]
A question out of curiosity, what prevents the `example.reference` to be a copy of `example.predicted` by default, so when you call `Example.from_dict(doc, annotations)`, it simply sets the annotations on top of what is already present in `example.reference`. 

I was hoping something like this should ideally work out of the box, but that isn't the case, and complaints when I try to access `doc.sents` in `my_component` later during training.:

```
with nlp.select_pipes(disable="my_component"):
    doc = nlp(text)
    yield Example.from_dict(doc, {'cats': {"my_label": label}})
```

where my pipe is `sentencizer >> my_component`.

**Reply by adrianeboyd** [reply]
I can see why that would be an attractive option in this case, but I think that in `Example.from_dict` we want to avoid having to deal with cases where both the predicted `doc` and the reference dict provide annotations for the same attribute.

Right now only the `words` from the predicted `doc` are used as part of the reference annotation if those are not provided in the reference dict and nothing else is shared.
